src/mutex.cc

```diff
@@ -21,27 +21,47 @@ namespace toit {
 void Locker::leave() {
   Thread* thread = Thread::current();
   if (thread->locker_ != this) FATAL("unlocking would break lock order");
-  thread->locker_ = previous_;
-  // Perform the actual unlock.
-  mutex_->unlock();
+  thread->locker_ = previous();
+  // Perform the actual unlock unless it was a reentrant locking.
+  bool unlock = !reentrant_;
+  reentrant_ = false;
```

Comment:
I think I would not reset this field.
If I'm not wrong the field should only be useful when the lock is taken.
You can then remove the `ASSERT` in line 42.